build-tools/packages/build-cli/src/commands/generate/sourceEntrypoints.ts

```diff
@@ -0,0 +1,386 @@
+	const rootDir = await readOutDirTsConfig(outDir);
+
+	const outFileSuffix = ".ts";
+
+	return new Map([
+		[`${rootDir}${ApiLevel.alpha}${outFileSuffix}`, ApiTag.alpha],
+		[`${rootDir}${ApiLevel.beta}${outFileSuffix}`, ApiTag.beta],
+		[`${rootDir}${ApiLevel.public}${outFileSuffix}`, ApiTag.public],
+		[`${rootDir}${ApiLevel.legacy}${outFileSuffix}`, ApiTag.legacy],
+	]);
+}
+
+/**
+ * Resolves a mapping of `ApiTag` levels to their modified export paths.
+ */
+function getOutputConfiguration(
+	mapApiTagToSourcePath: ReadonlyMap<string, ApiTag>,
+	mapApiTagToExportPath: Map<ApiTag, ExportData>,
+	logger: CommandLogger,
+): Map<ApiTag, ExportData> {
+	const result = new Map<ApiTag, ExportData>();
+	for (const [sourcePath, apiTag] of mapApiTagToSourcePath) {
+		// Check if apiTag exists in mapApiTagToOutputPath
+		const exportPath = mapApiTagToExportPath.get(apiTag);
+		if (exportPath) {
+			// Modify the exportPath
+			const modifiedExportPath = exportPath.relPath
+				.replace(/(lib|dist)/g, "src")
+				.replace(/\.js$|\.d\.ts$/, ".ts");
+
+			if (result.has(apiTag)) {
+				logger?.warning(`${modifiedExportPath} found in exports multiple times.`);
+			} else if (sourcePath === modifiedExportPath) {
+				result.set(apiTag, { ...exportPath, relPath: modifiedExportPath });
+			}
+		}
+	}
+
+	return result;
+}
+
+/**
+ * Read package "exports" to determine which "default"/ "types" paths to return along with `ApiTag`.
+ *
+ * @param packageJson - json content of package.json
+ * @param emitDeclarationOnly - If true, "types" exports are considered.
+ * @param logger - optional Logger
+ * @returns Map with API tags or levels with export path data
+ */
+function mapExportPathToApiTag(
+	packageJson: PackageJson,
+	tsconfig: TsConfigJson,
+	logger?: Logger,
+): Map<ApiTag, ExportData> {
+	const mapKeyToOutput = new Map<ApiTag, ExportData>();
+
+	let emitDeclarationOnly = false;
+	if (tsconfig.compilerOptions?.emitDeclarationOnly !== undefined) {
+		emitDeclarationOnly = tsconfig.compilerOptions.emitDeclarationOnly;
+	}
+
+	const { exports } = packageJson;
+
+	if (typeof exports !== "object" || exports === null) {
+		throw new Error('no valid "exports" within package properties');
+	}
+
+	// Iterate through exports looking for properties with values matching keys in map.
+	for (const [exportPath] of Object.entries(exports)) {
+		const resolvedExport = resolve.exports(packageJson, exportPath, {
+			conditions: emitDeclarationOnly ? ["types"] : ["default"],
+		});
+		if (resolvedExport === undefined || resolvedExport.length === 0) {
+			throw new Error(`exports for ${exportPath} is undefined`);
+		}
+
+		const level = exportPath === "." ? ApiTag.public : exportPath.replace("./", "");
+
+		if (!isKnownApiTag(level)) {
+			throw new Error(`${exportPath} is not a known API tag`);
+		}
+
+		if (level === ApiTag.internal) {
+			continue;
+		}
+
+		if (mapKeyToOutput.has(level)) {
+			logger?.warning(`${resolvedExport[0]} found in exports multiple times.`);
```

Comment:
I don't know how this version of things would ever get to this condition. Doesn't seem useful.